src/cpp/src/model_runner.hpp

```diff
@@ -129,12 +140,18 @@ class ModelRunner {
             for (size_t seq_id = 0; seq_id < num_running_sequences; ++seq_id) {
                 Sequence::CPtr sequence = running_sequences[seq_id];
 
-                for (size_t token_id = 0, position_id = group_position_id; token_id < num_scheduled_tokens; ++token_id, ++position_id) {
+                for (size_t token_id = 0, position_id = group_position_id; token_id < num_scheduled_tokens; ++token_id, ++position_id, ++gathering_current_index) {
                     // compute token for current sequence
                     input_ids_data[token_id] = position_id < sequence_group->get_prompt_len() ?
                         sequence_group->get_prompt_ids()[position_id] :
                         sequence->get_generated_ids()[position_id - sequence_group->get_prompt_len()];
 
+                    if (matmul_gathering_is_required) {
+                        if (group_position_id + token_id >= prompt_len - 1 || echo_output) {
```

Comment:
why not to use `requires_sampling()` for sequence group instead of this `group_position_id + token_id >= prompt_len - 1` condition? 

We can have situation when sequence group was computed, but later it was dropped because of memory requirements (e.g. high priority request has evicted low priority request). E.g. we have prompt 10 + 5 generated (sampled) tokens, when such sequence is evicted, then we don't need to sample all these 5 last tokens, and sampling is required only when we sample 16th token (while your condition does not imply it)